### PR TITLE
Simplify stack-to-register mapping API

### PR DESCRIPTION
**Pull request by guillep**
The compile-time symbolic stack should always be in sync with the (spilled) runtime stack.
If they get out of sync, the generated code is always wrong.

This PR fixes the API and implementation of stack mapping to prevent such desynchronisations, and provide a much simpler way to use.

### CODE REVIEW

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st [hunk 1]
```diff
@@ -134,6 +141,21 @@ CogSimStackEntry >> constant: anObject [
...
+	"Copy the value of the current entry to the register argument.
+	Do not touch the stack"
+	
+	type caseOf: {
+		[SSBaseOffset]	-> [cogit MoveMw: offset r: registerr R: reg].
+		[SSSpill]		-> [cogit MoveMw: offset r: registerr R: reg].
+		[SSConstant]	-> [cogit genMoveConstant: constant R: reg].
+		[SSRegister]	-> [reg ~= registerr
+								ifTrue: [cogit MoveR: registerr R: reg]
+								ifFalse: [cogit Label]] }
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/CogSimStackEntry.class.st]: Why a label is created if it's the same register? 
Can we change the condition to use `==` instead?

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st [hunk 2]
```diff
@@ -247,7 +255,7 @@ CogSimStackEntry >> moveToVectorReg: reg [
 { #category : #accessing }
 CogSimStackEntry >> offset [
 	"Answer the value of offset"
-	self assert: type = SSBaseOffset.
+	self assert: (self type == SSSpill or: [self type == SSBaseOffset]).
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/CogSimStackEntry.class.st:258]: Can we put a name to this?

File: smalltalksrc/VMMaker/VMMakerConfiguration.class.st [hunk 3]
```diff
@@ -160,19 +160,6 @@ VMMakerConfiguration class >> generateSqueakSpurCogVM [
...
-	"No primitives since we can use those for the Cog VM"
-	^VMMaker
-		generate: CoInterpreter
-		and: RegisterAllocatingCogit
-		with: #(ObjectMemory Spur32BitCoMemoryManager)
-		to: self sourceTree, '/spurregsrc'
-		platformDir: self sourceTree, '/platforms'
-		including:#()
-		configuration: self
-]
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/VMMakerConfiguration.class.st:174]: 💀

File: smalltalksrc/VMMakerTests/VMStackToRegisterMappingTest.class.st [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	
+	"The element should now be in a spilled position in the stack.
+	I.e., Relative to FPReg, with offset 4 (base 0) in the frame (the first three are saved FP, method, context, receiver)"
+	self assert: cogit ssTop type equals: SSSpill.
+	self assert: cogit ssTop base equals: FPReg.
+	self assert: cogit ssTop offset equals: -4 * memory wordSize.
+	
+	"But it is spilled"
+	self assert: cogit ssTop spilled
+]
```
Comment (PalumboN) [inline smalltalksrc/VMMakerTests/VMStackToRegisterMappingTest.class.st:243]: `Spilled = Flushed`?

**Review by PalumboN — APPROVED** [review]
RnR 🚀

File: smalltalksrc/VMMakerTests/VMStackToRegisterMappingTest.class.st (hunk 4 above)
Comment (guillep) [reply smalltalksrc/VMMakerTests/VMStackToRegisterMappingTest.class.st:243]: yes, I kept for now the old terminology, we can discuss it later.

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st (hunk 2 above)
Comment (guillep) [reply smalltalksrc/VMMaker/CogSimStackEntry.class.st:258]: we could, but it's only in two places, right? if we have a third one we abstract :)

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st (hunk 1 above)
Comment (guillep) [reply smalltalksrc/VMMaker/CogSimStackEntry.class.st]: > Why a label is created if it's the same register?

good question. I don't know. It was like that I just renamed this method.

> Can we change the condition to use `==` instead?

Make a suggestion and I accept it ;)

Comment (PalumboN) [reply smalltalksrc/VMMaker/CogSimStackEntry.class.st]: I didn't have the eggs to make the suggestion :P

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st (hunk 2 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/CogSimStackEntry.class.st:258]: Hahah three is the number... it's ok, only want to know why should be `SSSpill` or `SSBaseOffset` (for someone that doesn't know the possible types as me)

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st (hunk 1 above)
Comment (guillep) [reply smalltalksrc/VMMaker/CogSimStackEntry.class.st]:
```suggestion
		[SSRegister]	-> [reg = registerr
								ifTrue: [cogit Label]
								ifFalse: [cogit MoveR: registerr R: reg]] }
```

File: smalltalksrc/VMMaker/CogSimStackEntry.class.st (hunk 2 above)
Comment (guillep) [reply smalltalksrc/VMMaker/CogSimStackEntry.class.st:258]: Actually, both spill and baseoffset represent memory addresses in the form base+offset.
The only difference I've seen so far is that base+offset points to any potential address (e.g, object+variable offset), while spill is always a spill slot in the frame.

But we could differentiate one from the other using the spilled variable 🤷

**Pull request merged**